File: services/shared/httpx/response.go
package httpx

import (
	"encoding/json"
	"net/http"
	"time"
)

type APIResponse struct {
	Timestamp time.Time   `json:"timestamp"`
	Status    int         `json:"status"`
	Success   bool        `json:"success"`
	Message   string      `json:"message,omitempty"`
	Data      interface{} `json:"data,omitempty"`
	Error     interface{} `json:"error,omitempty"`
}

func WriteJSON(w http.ResponseWriter, status int, success bool, message string, data interface{}, errPayload interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(APIResponse{
		Timestamp: time.Now().UTC(),
		Status:    status,
		Success:   success,
		Message:   message,
		Data:      data,
		Error:     errPayload,
	})
}

func OK(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusOK, true, message, data, nil)
}

func Created(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusCreated, true, message, data, nil)
}

func Error(w http.ResponseWriter, status int, message string, err error) {
	payload := map[string]string{"message": message}
	if err != nil {
		payload["detail"] = err.Error()
	}
	WriteJSON(w, status, false, message, nil, payload)
}

// BadRequest sends a 400 Bad Request response
func BadRequest(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusBadRequest, false, message, data, nil)
}

// InternalServerError sends a 500 Internal Server Error response
func InternalServerError(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusInternalServerError, false, message, data, nil)
}

// NotFound sends a 404 Not Found response
func NotFound(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusNotFound, false, message, data, nil)
}

// Unauthorized sends a 401 Unauthorized response
func Unauthorized(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusUnauthorized, false, message, data, nil)
}

// Forbidden sends a 403 Forbidden response
func Forbidden(w http.ResponseWriter, message string, data interface{}) {
	WriteJSON(w, http.StatusForbidden, false, message, data, nil)
}

File: services/blog-service/src/utils/response.ts
import { FastifyReply } from 'fastify';

/**
 * Standardized API response structure matching Go services
 */
export interface APIResponse {
  timestamp: string;
  status: number;
  success: boolean;
  message?: string;
  data?: any;
  error?: any;
}

/**
 * Response utility class for consistent API responses
 */
export class ResponseHelper {
  /**
   * Send a successful response
   */
  static ok(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 200,
      success: true,
      message,
      data,
    };
    reply.status(200).send(response);
  }

  /**
   * Send a created response
   */
  static created(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 201,
      success: true,
      message,
      data,
    };
    reply.status(201).send(response);
  }

  /**
   * Send an error response
   */
  static error(reply: FastifyReply, status: number, message: string, error?: any): void {
    const errorPayload: any = { message };
    if (error) {
      errorPayload.detail = error instanceof Error ? error.message : error;
    }

    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status,
      success: false,
      message,
      error: errorPayload,
    };
    reply.status(status).send(response);
  }

  /**
   * Send a bad request response (400)
   */
  static badRequest(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 400,
      success: false,
      message,
      error: data,
    };
    reply.status(400).send(response);
  }

  /**
   * Send a not found response (404)
   */
  static notFound(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 404,
      success: false,
      message,
      error: data,
    };
    reply.status(404).send(response);
  }

  /**
   * Send an unauthorized response (401)
   */
  static unauthorized(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 401,
      success: false,
      message,
      error: data,
    };
    reply.status(401).send(response);
  }

  /**
   * Send a forbidden response (403)
   */
  static forbidden(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 403,
      success: false,
      message,
      error: data,
    };
    reply.status(403).send(response);
  }

  /**
   * Send an internal server error response (500)
   */
  static internalServerError(reply: FastifyReply, message: string, data?: any): void {
    const response: APIResponse = {
      timestamp: new Date().toISOString(),
      status: 500,
      success: false,
      message,
      error: data,
    };
    reply.status(500).send(response);
  }
}
